salt/modules/zfs.py

```diff
@@ -362,164 +293,229 @@ def list_(name=None, **kwargs):
         display numbers in parsable (exact) values
         .. versionadded:: Oxygen
 
+    .. versionadded:: 2015.5.0
+
     CLI Example:
 
     .. code-block:: bash
 
         salt '*' zfs.list
         salt '*' zfs.list myzpool/mydataset [recursive=True|False]
         salt '*' zfs.list myzpool/mydataset properties="sharenfs,mountpoint"
+
     '''
     ret = OrderedDict()
-    zfs = _check_zfs()
-    recursive = kwargs.get('recursive', False)
-    depth = kwargs.get('depth', 0)
+
+    ## update properties
+    # NOTE: properties should be a list
     properties = kwargs.get('properties', 'used,avail,refer,mountpoint')
-    sort = kwargs.get('sort', None)
-    ltype = kwargs.get('type', None)
-    order = kwargs.get('order', 'ascending')
-    parsable = kwargs.get('parsable', False)
-    cmd = '{0} list -H'.format(zfs)
-
-    # parsable output
-    if parsable:
-        cmd = '{0} -p'.format(cmd)
-
-    # filter on type
-    if ltype:
-        cmd = '{0} -t {1}'.format(cmd, ltype)
-
-    # recursively list
-    if recursive:
-        cmd = '{0} -r'.format(cmd)
-        if depth:
-            cmd = '{0} -d {1}'.format(cmd, depth)
-
-    # add properties
-    properties = properties.split(',')
-    if 'name' in properties:  # ensure name is first property
+    if not isinstance(properties, list):
+        properties = properties.split(',')
+
+    # NOTE: name should be first property
+    while 'name' in properties:
```

Comment:
Because it is possible to enter 'name' multiple times. So we remove all of them.

```
Python 2.7.10 (default, Jul 15 2017, 17:16:57)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> test = ['hello', 'name', 'world', 'name', 'cookies']
>>> if 'name' in test:
...     test.remove('name')
...
>>> print test
['hello', 'world', 'name', 'cookies']
>>> test = ['hello', 'name', 'world', 'name', 'cookies']
>>> while 'name' in test:
...     test.remove('name')
...
>>> print test
['hello', 'world', 'cookies']
>>>
```

I have update the note to make this clear why we do this.